celery/app/trace.py

```diff
@@ -369,7 +382,12 @@ def build_tracer(name, task, loader=None, hostname=None, store_errors=True,
     from celery import canvas
     signature = canvas.maybe_signature  # maybe_ does not clone if already
 
-    def on_error(request, exc, state=FAILURE, call_errbacks=True):
+    def on_error(
+            request: celery.app.task.Context,
+            exc: Union[Exception, Type[Exception]],
+            state: str = FAILURE,
+            call_errbacks: bool = True) -> Tuple[Info, Any, Any, Any]:
```

Comment:
I believe that `I.state` is a string, but that can always be tightened up in the future.

```suggestion
            call_errbacks: bool = True) -> Tuple[Info, Any, str, Any]:
```